### PR TITLE
fix(api): added functionality to allow sending auth token alongside action queue

### PR DESCRIPTION
**Pull request by cccs-mdr**
This PR fixes custom actions to have access to the user's auth token again, since actions no longer run inside a flask request if the action queue is enabled.

---

This pull request introduces secure encryption and decryption of authorization tokens for queued actions, adds configuration and validation for the encryption key, and updates the action execution flow to handle tokens securely. It also includes improvements to logging and test coverage for these changes.

**Secure Authorization Token Handling:**

- Added support for encrypting and decrypting authorization tokens (both JWT and non-JWT) before storing them in Redis queues, using either JWE (for JWTs) or AES-GCM (for non-JWTs). This ensures sensitive credentials are not stored in plaintext and are handled securely throughout the action queue workflow. (`api/howler/services/auth_service.py` [[1]](diffhunk://#diff-7ce9d12525c3d6eda0534181ae4994f0aeb2b3c2f2241348bda853afaf5041d1R4-R9) [[2]](diffhunk://#diff-7ce9d12525c3d6eda0534181ae4994f0aeb2b3c2f2241348bda853afaf5041d1R45-R93); `api/howler/services/jwt_service.py` [[3]](diffhunk://#diff-8fab217af95c74889c9982e438d92ca4de39f9d8ff3170da5f725e40c52960f7R9) [[4]](diffhunk://#diff-8fab217af95c74889c9982e438d92ca4de39f9d8ff3170da5f725e40c52960f7R18-R45)
- Modified the action queue and batch processing logic to include encrypted `auth_token` in the queued action payload, and to decrypt it before executing actions. (`api/howler/services/action_service.py` [[1]](diffhunk://#diff-5f532aa591c4c6b094fee91c3cdc3519df26b9196410901caeca9a7e63c22b64R30-R53) [[2]](diffhunk://#diff-5f532aa591c4c6b094fee91c3cdc3519df26b9196410901caeca9a7e63c22b64R88-R125) [[3]](diffhunk://#diff-5f532aa591c4c6b094fee91c3cdc3519df26b9196410901caeca9a7e63c22b64L102-R160)
- Updated the action execution logic to pass the decrypted `auth_token` to external operations, ensuring downstream services can authenticate requests. (`api/howler/services/action_service.py` [api/howler/services/action_service.pyR244-R247](diffhunk://#diff-5f532aa591c4c6b094fee91c3cdc3519df26b9196410901caeca9a7e63c22b64R244-R247))

**Configuration and Validation:**

- Introduced a new `jwe_secret_key` field in the system configuration, with a validator to ensure it is exactly 32 bytes (UTF-8 encoded), as required for AES-GCM encryption. (`api/howler/odm/models/config.py` [[1]](diffhunk://#diff-52ee7c226148cc8d84fc36b1a9b9dec6db434703478e48aca8a8fd2ce04b392fR479-R482) [[2]](diffhunk://#diff-52ee7c226148cc8d84fc36b1a9b9dec6db434703478e48aca8a8fd2ce04b392fR492-R499)
- Updated test configuration to provide a valid `jwe_secret_key`. (`api/test/unit/config.yml` [api/test/unit/config.ymlR38](diffhunk://#diff-de298c8fb55c0421b248413612a984f59f69f2523524f05cbf0cc2e20f0fb799R38))

**API and Usability Improvements:**

- Allowed `enqueue_action_execution` to accept either a single hit ID or a list of hit IDs, simplifying the API and reducing boilerplate. (`api/howler/services/action_service.py` [[1]](diffhunk://#diff-5f532aa591c4c6b094fee91c3cdc3519df26b9196410901caeca9a7e63c22b64L54-R74); `api/howler/services/hit_service.py` [[2]](diffhunk://#diff-4329e287b29f436c78c60441e0b589847e05e0d428024ff7b3c2231c72553d95L730-R730)
- Improved logging for action queue processing and hit transitions for better traceability and debugging. (`api/howler/cronjobs/action_queue_worker.py` [[1]](diffhunk://#diff-5dac40fa52b7cc9677258462307ebce95d7ee467e7dabfe0b8189f4b0da9009bL59-R64); `api/howler/services/hit_service.py` [[2]](diffhunk://#diff-4329e287b29f436c78c60441e0b589847e05e0d428024ff7b3c2231c72553d95L687-R687) [[3]](diffhunk://#diff-4329e287b29f436c78c60441e0b589847e05e0d428024ff7b3c2231c72553d95R739-R741)

**Testing and Validation:**

- Added comprehensive unit tests to verify correct encryption and decryption of tokens, proper delegation for JWT tokens, and handling of edge cases (e.g., missing tokens, non-JWT tokens with similar structure). (`api/test/unit/services/test_auth_service.py` [api/test/unit/services/test_auth_service.pyR87-R130](diffhunk://#diff-ad8bc1c2d51fb150cc39315e294308ebeb9328c9356d7a5c3aab4d91227b325dR87-R130))
- Updated test utilities and dependencies to support new configuration and service logic. (`api/test/unit/services/test_enqueue_action.py` [api/test/unit/services/test_enqueue_action.pyR1-R8](diffhunk://#diff-c14f566e8efa12fde5a24351116fa5533319c85d455b8e6e634fafb12123d9b5R1-R8))

**Other:**

- Bumped the API version to reflect these significant security and API changes. (`api/pyproject.toml` [api/pyproject.tomlL155-R155](diffhunk://#diff-40f8f985c548e2b57b803d17f42833f8a4d94603ca5b5e0bf48de51872809400L155-R155))

These changes collectively enhance the security, reliability, and maintainability of Howler's action queue and authorization handling.

### CODE REVIEW

File: api/howler/services/action_service.py [hunk 1]
```diff
@@ -195,6 +241,10 @@ def bulk_execute_on_query(query: str, trigger: str = "create", user: Optional[Us
             if not user:
                 raise NotImplementedError("Running actions without a user object is not currently supported")
 
+            if not _is_builtin_operation(operation.operation_id):
+                logger.warning("Appending auth token to external operation %s", operation.operation_id)
+                parsed_data["auth_token"] = auth_token
```
[the remark this answers is not in the document]
Comment (alexandre-gauvin) [reply api/howler/services/action_service.py]: I would agree with this one, I see that you seem to have verify its state in other functions using it but adding the safety of not making it null could help

File: api/howler/services/hit_service.py [hunk 2]
```diff
@@ -684,7 +684,7 @@ def transition_hit(
     workflow: Workflow = get_hit_workflow()
 
     # Log hit that will be transitioned
-    logger.debug("Transitioning (%s)", hit)
+    logger.debug("Transitioning hit %s with transition %s", hit["howler"]["id"])
```
[the remark this answers is not in the document]
Comment (alexandre-gauvin) [reply api/howler/services/hit_service.py]: Yeah, this need to be acted on. This will create an exception every time its passed on

**Review by alexandre-gauvin — CHANGES_REQUESTED** [review]
Everything seems good except the two thing Copilot mentioned. 

The auth one would still work since how we normally work with most api and you updated the function to be able to handle it. But adding safety wouldn't be too bad


The %s I tough it worked in python but after verification it is creating error.

**Comment by alexandre-gauvin** [comment]
Ah I wasn't updated I'll verify these section back and approve

**Review by alexandre-gauvin — APPROVED** [review]

File: api/howler/services/action_service.py [hunk 3]
```diff
@@ -25,12 +27,30 @@ class TriggeredAction(TypedDict):
...
+    return find_spec(f"howler.actions.{operation_id}") is not None
+
+
+def _request_auth_token() -> str | None:
+    """Extract the credential portion of the current request's authorization header."""
+    if not has_request_context():
+        return None
+
+    authorization = request.headers.get("Authorization")
+    if authorization is None or " " not in authorization:
```
Comment (cccs-tsha) [inline api/howler/services/action_service.py:48]: Question for this, in this situation would it be possible to have a case where `bearer abc123 456` occurs? If so, will that be resolved to true?

File: api/howler/services/action_service.py [hunk 4]
```diff
@@ -65,25 +85,44 @@ def enqueue_action_execution(hit_ids: list[str], trigger: str = "create", user:
     if not hit_ids:
         return
 
+    if not isinstance(hit_ids, list):
```
Comment (cccs-tsha) [inline api/howler/services/action_service.py:88]: Similar question here, there is no case in which the hit_ids can be multiple but in 1 string? Ex. "hit1, hit2, hit3..."

File: api/howler/services/jwt_service.py [hunk 5]
```diff
@@ -14,6 +15,34 @@
...
+_JWE_HEADER = {"alg": "A256GCMKW", "enc": "A256GCM"}
+
+
+def _jwe_secret() -> bytes:
+    """Return the configured encryption key for JWE payloads."""
+    secret_key = config.system.jwe_secret_key
+    if secret_key is None:
+        raise HowlerValueError("System jwe_secret_key must be configured before encrypting tokens")
+
+    return secret_key.encode("utf-8")
```
Comment (cccs-tsha) [inline api/howler/services/jwt_service.py]: This entire function also exists within auth_service with a different name, `_token_encryption_key` with the same value error message being thrown. Maybe we could turn this into a util of some sort for reusability?

**Review by cccs-tsha — APPROVED** [review]
LGTM! I just had a couple questions and a suggestion, feel free to take a look if you think it makes sense :)

**Review by brtobal — APPROVED** [review]
Looks good to me

File: api/howler/services/action_service.py (hunk 3 above)
Comment (cccs-mdr) [reply api/howler/services/action_service.py:48]: It would be malformed, so I don't think we need to cover it.

File: api/howler/services/action_service.py (hunk 4 above)
Comment (cccs-mdr) [reply api/howler/services/action_service.py:88]: Since this is an internal service, no. We validate the ids before they are sent here.

File: api/howler/services/jwt_service.py (hunk 5 above)
Comment (cccs-mdr) [reply api/howler/services/jwt_service.py]: Good point, modified!
